crates/matrix-sdk-ui/src/spaces/mod.rs

```diff
@@ -0,0 +1,510 @@
+// Copyright 2025 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for that specific language governing permissions and
+// limitations under the License.
+
+//! High level interfaces for working with Spaces
+//!
+//! The `SpaceService` is an UI oriented, high-level interface for working with
+//! Matrix Spaces. It provides methods to retrieve joined spaces, subscribe
+//! to updates, and navigate space hierarchies.
+//!
+//! It consists of 3 main components:
+//! - `SpaceService`: The main service for managing spaces. It
+//! - `SpaceGraph`: An utility that maps the `m.space.parent` and
+//!   `m.space.child` fields into a graph structure, removing cycles and
+//!   providing access to top level parents.
+//! - `SpaceRoomList`: A component for retrieving a space's children rooms and
+//!   their details.
+
+use std::sync::Arc;
+
+use eyeball_im::{ObservableVector, VectorSubscriberBatchedStream};
+use futures_util::pin_mut;
+use imbl::Vector;
+use matrix_sdk::{
+    Client, deserialized_responses::SyncOrStrippedState, executor::AbortOnDrop, locks::Mutex,
+};
+use matrix_sdk_common::executor::spawn;
+use ruma::{
+    OwnedRoomId,
+    events::{
+        SyncStateEvent,
+        space::{child::SpaceChildEventContent, parent::SpaceParentEventContent},
+    },
+};
+use tracing::error;
+
+use crate::spaces::graph::SpaceGraph;
+pub use crate::spaces::{room::SpaceRoom, room_list::SpaceRoomList};
+
+pub mod graph;
+pub mod room;
+pub mod room_list;
+
+/// The main entry point into the Spaces facilities.
+///
+/// The spaces service is responsible for retrieving one's joined rooms,
+/// building a graph out of their `m.space.parent` and `m.space.child` state
+/// events, and providing access to the top-level spaces and their children.
+///
+/// # Examples
+///
+/// ```no_run
+/// use futures_util::StreamExt;
+/// use matrix_sdk::Client;
+/// use matrix_sdk_ui::spaces::SpaceService;
+/// use ruma::owned_room_id;
+///
+/// # futures_executor::block_on(async {
```

Comment:
Oh, it's for this? Since the example is `no_run` you don't really need this, having the `async` block is enough, for example like here: https://github.com/matrix-org/matrix-rust-sdk/blob/3aeface647183345973e310930fa037bc603db04/crates/matrix-sdk-ui/src/sync_service.rs#L572-L603.

Reply:
Ah sorry, this was supposed to work without `no_run` through that `futures-executor` but it ends up just failing miseraby with something along the lines of:

```
% cargo test spaces -p matrix-sdk-ui --doc

test crates/matrix-sdk-ui/src/spaces/mod.rs - spaces::SpaceService (line 60) ... FAILED

failures:

---- crates/matrix-sdk-ui/src/spaces/mod.rs - spaces::SpaceService (line 60) stdout ----
Test executable failed (exit status: 101).

stderr:

thread 'main' (14505151) panicked at /var/folders/xy/jh61h2x979g3r7c5rglgd_800000gn/T/rustdoctestbJlRWS/doctest_bundle_2024.rs:48:22:
not yet implemented
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
We discussed this in chat. Using `no_run` is completely fine, we don't want our examples to try to reach real homeservers anyways.